### PR TITLE
Implement float->int trunc instructions

### PR DESCRIPTION
**Pull request by chfast**
The operator from the wasm spec is partial, i.e. it has some undefined results.
image
But, the execution spec requires to trap on operator's undefined result.
image

### CODE REVIEW

**Comment by gumb0** [comment]
Please rebase.

File: test/unittests/execute_floating_point_test.cpp [hunk 1]
```diff
@@ -69,4 +70,174 @@ TEST(execute_floating_point, f64_add)
...
+        const auto arg =
+            std::nextafter(std::pow(src_type{2}, src_type{dst_limits::digits}), src_type{0});
+        const auto expected = static_cast<dst_type>(arg);
+        const auto result = execute(*instance, 0, {arg});
+        EXPECT_EQ(result.value.template as<dst_type>(), expected);
+    }
+
+    {  // Min defined: -2^N or 0.
+        const auto arg =
+            dst_limits::is_signed ? -std::pow(src_type{2}, src_type{dst_limits::digits}) : 0;
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:206]: Maybe would make sense to test 0 for signed types, too, and move `-2^N` under `if (dst_limits::is_signed)` at the bottom of the test.

**Comment by axic** [comment]
How come this still does not affect spec tests expectations?

File: test/unittests/execute_floating_point_test.cpp (hunk 1 above)
Comment (chfast) [reply test/unittests/execute_floating_point_test.cpp]: Done.

File: lib/fizzy/execute.cpp [hunk 2]
```diff
@@ -1375,6 +1390,42 @@ ExecutionResult execute(Instance& instance, FuncIdx func_idx, span<const Value>
             stack.push(static_cast<uint32_t>(stack.pop()));
             break;
         }
+        case Instr::i32_trunc_f32_s:
+        {
+            if (!trunc<float, int32_t>(stack))
+            {
+                trap = true;
```
Comment (gumb0) [inline lib/fizzy/execute.cpp:1397]: Do we trap instead of undefined result? Why? This seems against the spec.

File: lib/fizzy/execute.cpp [hunk 3]
```diff
@@ -324,6 +326,19 @@ inline DstT extend(SrcT in) noexcept
         return DstT{in};
 }
 
+template <typename SrcT, typename DstT>
+inline bool trunc(OperandStack& stack) noexcept
+{
+    using boundaries = trunc_boundaries<SrcT, DstT>;
```
Comment (gumb0) [inline lib/fizzy/execute.cpp:332]: Why would it not work with `static_cast<SrcT>(std::numeric_limits<Dst>::min())` / `static_cast<SrcT>(std::numeric_limits<Dst>::max())` ?

File: lib/fizzy/execute.cpp (hunk 2 above)
Comment (chfast) [reply lib/fizzy/execute.cpp:1397]: I updated the description with the explanation. And spec tests confirm that.

File: lib/fizzy/execute.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/execute.cpp:332]: E.g.
- `-0.0 -> 0u`
- `-0.999 -> 0u`
- `-1.0 -> undefined`

File: test/unittests/execute_floating_point_test.cpp [hunk 4]
```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
...
+        // The lower boundary is:
+        // - for signed integers: -2^N - 1,
+        // - for unsigned integers: -1.
+        // However, the -2^N - 1 may be not representative in a float type so we compute it as
+        // floor(-2^N - epsilon).
+        const auto min_defined_int = IntLimits::min();
+        const auto lower_boundary =
+            std::floor(std::nextafter(FloatT{min_defined_int}, -FloatLimits::infinity()));
+        EXPECT_EQ(lower_boundary, expected_boundaries::lower);
+        EXPECT_THAT(execute(*instance, 0, {upper_boundary}), Traps());
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:240]:
```suggestion
        EXPECT_THAT(execute(*instance, 0, {lower_boundary}), Traps());
```

File: test/unittests/execute_floating_point_test.cpp [hunk 5]
```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
...
+    {
+        // BOUNDARIES OF DEFINITION
+        //
+        // Here we want to identify and test the boundary values of the defined behavior of the
+        // trunc instructions. For undefined results the execution must trap.
+        // Note that floating point type can represent any power of 2.
+
+        using expected_boundaries = trunc_boundaries<FloatT, IntT>;
+
+        // For iN with max value 2^N-1 the float(2^N) exists and trunc() from that value to iN
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:218]: To disambiguate from "max value" in the beginning of the sentence
```suggestion
        // For iN with max value 2^N-1 the float(2^N) exists and trunc(float(2^N)) to iN
```

File: test/unittests/execute_floating_point_test.cpp [hunk 6]
```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
...
+        // is undefined.
+        const auto upper_boundary = std::pow(FloatT{2}, FloatT{IntLimits::digits});
+        EXPECT_EQ(upper_boundary, expected_boundaries::upper);
+        EXPECT_THAT(execute(*instance, 0, {upper_boundary}), Traps());
+
+        // But the trunc() of the next float value smaller than 2^N is defined.
+        // Depending on the resolution of the floating point type, the result integer value may
+        // be other than 2^(N-1).
+        const auto max_defined = std::nextafter(upper_boundary, FloatT{0});
+        const auto max_defined_int = static_cast<IntT>(max_defined);
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:228]: This would be caught by UBSan in case the assumption that it's defined were wrong?

File: lib/fizzy/trunc_boundaries.hpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace fizzy
+{
+template <typename SrcT, typename Dst>
+struct trunc_boundaries;
+
+template <>
+struct trunc_boundaries<float, int32_t>
+{
+    static constexpr float lower = -2147483904.0;
```
[the remark this answers is not in the document]
Comment (gumb0) [reply lib/fizzy/trunc_boundaries.hpp:17]: I see explanation in the test

[the remark this answers is not in the document]
Comment (chfast) [reply lib/fizzy/trunc_boundaries.hpp:17]: At this range `float` has big gaps in values. This is max value being smaller than `-2^31`.

File: test/unittests/execute_floating_point_test.cpp (hunk 6 above)
Comment (chfast) [reply test/unittests/execute_floating_point_test.cpp:228]: Yes, UBSan catches the results that are undefined in wasm (because this is undefined behavior in C/C++).

File: lib/fizzy/execute.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/execute.cpp:332]: I think it is rather difficult to have generic formula for upper/lower boundaries. However, it is possible to compress these (e.g. for unsigned type these are `-1.0` and `2^N`. But I decided to leave these expanded at least until we have confirmation that all spectests are passed.

Having separate file for it is not ideal, but I wanted to check the values against math formulas used in unit tests.

File: lib/fizzy/trunc_boundaries.hpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cstdint>
+
+namespace fizzy
+{
+template <typename SrcT, typename Dst>
+struct trunc_boundaries;
```
[the remark this answers is not in the document]
Comment (chfast) [reply lib/fizzy/trunc_boundaries.hpp:14]: Comment added.

File: test/unittests/execute_floating_point_test.cpp [hunk 9]
```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
...
+        EXPECT_THAT(execute(*instance, 0, {max_defined}), Result(max_defined_int));
+
+        // The lower boundary is:
+        // - for signed integers: -2^N - 1,
+        // - for unsigned integers: -1.
+        // However, the -2^N - 1 may be not representative in a float type so we compute it as
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:234]: I thnk this explanation how `lower_bound` is calculated should be in `trunc_boundaries.hpp`, too.

File: lib/fizzy/execute.cpp (hunk 3 above)
Comment (gumb0) [reply lib/fizzy/execute.cpp:332]: I think it looks fine now, I like that it's quite simple. The only thing is that  bounds look like magic values, they are not defined in spec. It needs a comment to explain where they come from.

File: test/unittests/execute_floating_point_test.cpp [hunk 10]
```diff
@@ -66,4 +69,203 @@ TEST(execute_floating_point, f64_add)
...
+
+        const auto min_defined = std::nextafter(lower_boundary, FloatT{0});
+        EXPECT_THAT(execute(*instance, 0, {min_defined}), Result(min_defined_int));
+    }
+
+    {
+        // NaNs.
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::quiet_NaN()}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {FloatLimits::signaling_NaN()}), Traps());
+        EXPECT_THAT(execute(*instance, 0, {nan<FloatT>()}), Traps());
```
Comment (gumb0) [inline test/unittests/execute_floating_point_test.cpp:250]: Can there be a NaN with sign bit set (`-nan<FloatT<()`)?

File: lib/fizzy/execute.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/execute.cpp:332]: Added comments.

File: test/unittests/execute_floating_point_test.cpp (hunk 10 above)
Comment (chfast) [reply test/unittests/execute_floating_point_test.cpp:250]: Yes, it can be.

Comment (chfast) [reply test/unittests/execute_floating_point_test.cpp:250]: Added test cases for these.

File: circle.yml [hunk 11]
```diff
@@ -271,9 +271,9 @@ jobs:
       - benchmark:
           min_time: "0.01"
       - spectest:
-          expected_passed: 5581
-          expected_failed: 9
-          expected_skipped: 6223
+          expected_passed: 5585
+          expected_failed: 13
```
Comment (axic) [inline circle.yml:275]: What are the new failed ones?

Comment (chfast) [reply circle.yml:275]: Should not be, I must update this.

File: test/unittests/execute_floating_point_test.cpp (hunk 9 above)
Comment (chfast) [reply test/unittests/execute_floating_point_test.cpp:234]: Added documentation.

File: lib/fizzy/execute.cpp [hunk 12]
```diff
@@ -324,6 +326,19 @@ inline DstT extend(SrcT in) noexcept
...
+
+    const auto in = stack.top().as<SrcT>();
```
Comment (axic) [inline lib/fizzy/execute.cpp:334]:
```suggestion
    const auto input = stack.top().as<SrcT>();
```
?

File: test/unittests/execute_floating_point_test.cpp [hunk 13]
```diff
@@ -66,4 +69,209 @@ TEST(execute_floating_point, f64_add)
...
+    (func (param f32) (result i32)
+      local.get 0
+      i32.trunc_f32_s
+    )
+    */
+    auto wasm = from_hex("0061736d0100000001060160017d017f030201000a070105002000a80b");
+
+    // Find and replace changeable values: types and the conversion instruction.
+    auto& param_type = wasm[wasm.find(static_cast<uint8_t>(ValType::f32))];
+    auto& result_type = wasm[wasm.find(static_cast<uint8_t>(ValType::i32))];
```
Comment (axic) [inline test/unittests/execute_floating_point_test.cpp:187]: Perhaps adding assertion that there are only a single occurrence of these valtypes?

File: lib/fizzy/trunc_boundaries.hpp [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+#pragma once
+
+#include <cstdint>
+
+namespace fizzy
+{
+/// Provides _exclusive_ lower and upper boundaries of the range
+/// for which SrcT -> DstT trunc operation has defined results.
+/// Theoretical values are (INTEGER_MIN - 1, INTEGER_MAX + 1) what is
```
Comment (gumb0) [inline lib/fizzy/trunc_boundaries.hpp:13]:
```suggestion
/// Theoretical values are (INTEGER_MIN - 1, INTEGER_MAX + 1) which is
```

File: lib/fizzy/execute.cpp [hunk 15]
```diff
@@ -324,6 +326,19 @@ inline DstT extend(SrcT in) noexcept
         return DstT{in};
 }
 
+template <typename SrcT, typename DstT>
+inline bool trunc(OperandStack& stack) noexcept
+{
+    using boundaries = trunc_boundaries<SrcT, DstT>;
+
+    const auto input = stack.top().as<SrcT>();
+    if (std::isnan(input) || input <= boundaries::lower || input >= boundaries::upper)
```
Comment (axic) [inline lib/fizzy/execute.cpp:335]: Why doesn't this need to check for `inf`/`-inf`?

Comment (chfast) [reply lib/fizzy/execute.cpp:335]: Because always: `-inf <= boundaries::lower` and `inf >= boundaries::upper`.

Comment (chfast) [reply lib/fizzy/execute.cpp:335]: Oh, actually, this can be optimized :/

Comment (axic) [reply lib/fizzy/execute.cpp:335]: Even if that is true (seemingly it is because the tests pass), we should have a static_assert here for that.

Comment (chfast) [reply lib/fizzy/execute.cpp]: I'm not sure what assert you would like to have here.

Comment (axic) [reply lib/fizzy/execute.cpp]:
```
-inf <= boundaries::lower
inf >= boundaries::lower
```

File: lib/fizzy/execute.cpp [hunk 16]
```diff
@@ -324,6 +326,20 @@ inline DstT extend(SrcT in) noexcept
...
+{
+    using boundaries = trunc_boundaries<SrcT, DstT>;
+
+    const auto input = stack.top().as<SrcT>();
+    if (input > boundaries::lower && input < boundaries::upper)
+    {
+        stack.top() = static_cast<DstT>(input);
+        return true;
+    }
+    return false;
```
Comment (axic) [inline lib/fizzy/execute.cpp:340]: After the optimisation I'd add `assert(std::isnan(input) || input == inf || input == -inf);` so that it matches the spec.

Comment (chfast) [reply lib/fizzy/execute.cpp:343]: Ok, added correct variants of these asserts.

Comment (axic) [reply lib/fizzy/execute.cpp:343]: Don't want to stretch this too much, but can we also have the static_assert for inf vs. boundaries? Obviously we cannot make the nan-check compile time :)

File: lib/fizzy/execute.cpp (hunk 15 above)
Comment (chfast) [reply lib/fizzy/execute.cpp]: No, this does not make sense to me. Every value is `>= -inf`.

**Comment by chfast** [comment]
> Looks good to me otherwise.

So what now?

**Comment by axic** [comment]
> So what now?

Dunno, merge?
